File: Math/Factorial.java
/*
What is a factorial of a number?
The factorial of a non-negative integer n is the product of all positive integers less than or equal to n. It is denoted by n!. For example, the factorial of 5 is 5! = 5 x 4 x 3 x 2 x 1 = 120.

Approach:
The recursive approach is used to calculate the factorial. The function `factorial()` takes an integer `n` as input. In each recursive call, we check if `n` is either 0 or 1 (base case), and if so, we return 1. Otherwise, we multiply `n` with the factorial of `n-1` and return the result.

Time Complexity: O(n)
The time complexity of this algorithm is O(n) because the recursion depth is equal to n, and in each recursion, we perform a constant-time operation.

Space Complexity: O(n)
The space complexity is O(n) because, in the worst-case scenario, the recursion depth can reach up to n. So, n frames will be added to the call stack.

Sample Input: 5
Sample Output: Factorial of 5 is: 120

*/

public class Factorial {
  // Recursive function to calculate the factorial
  public static int factorial(int n) {
    // Base case: if n is 0 or 1, the factorial is always 1
    if (n == 0 || n == 1) {
      return 1;
    }
    
    // Recursive case: multiply n with factorial of (n-1)
    return n * factorial(n - 1);
  }

  public static void main(String[] args) {
    int number = 5; // Sample input: calculate factorial of 5
    int result = factorial(number);
    System.out.println("Factorial of " + number + " is: " + result);
  }
}

File: Math/Factorial.cpp
/*
Explanation:
- The `factorial` function takes an integer `num` as input and calculates its factorial. The factorial of a non-negative integer N is the product of all positive integers less than or equal to N.
- In the function, we have a base case that handles the special case when `num` is 0. The factorial of 0 is defined as 1, so we return 1 for `num` equal to 0.
- For non-zero `num`, we use recursion to calculate the factorial. We call the `factorial` function with `num - 1`, and then multiply the result by `num`.
- The recursion continues until we reach the base case when `num` becomes 0, and the recursive calls start to return their values and calculate the final factorial.

Time Complexity: The time complexity of the `factorial` function is O(N), where N is the value of the input number. This is because the function makes N recursive calls to calculate the factorial.
  
Space Complexity: The space complexity is O(N) due to the recursion stack space used during the recursive calls. Each recursive call adds a new stack frame, and in the worst case, there will be N stack frames corresponding to the N recursive calls.

Sample Input:
Number: 5

Sample Output:
Factorial of 5 is 120
*/

#include <iostream>

// Function to calculate factorial
unsigned long long factorial(int num) {
    // Base case: factorial of 0 is 1
    if (num == 0) {
        return 1;
    }

    // Recursive case: calculate factorial using recursion
    return num * factorial(num - 1);
}

int main() {
    // Sample input
    int num = 5;

    // Calculate the factorial of the number
    unsigned long long result = factorial(num);

    // Print the result
    std::cout << "Factorial of " << num << " is " << result << std::endl;

    return 0;
}
